File: Exámenes/2016 SEPT/Intento 1/prHuffman/Huffman.java

/**
 * Huffman trees and codes.
 *
 * Data Structures, Grado en Informatica. UMA.
 *
 *
 * Student's name:
 * Student's group:
 */

import dataStructures.dictionary.AVLDictionary;
import dataStructures.dictionary.Dictionary;
import dataStructures.list.LinkedList;
import dataStructures.list.List;
import dataStructures.priorityQueue.BinaryHeapPriorityQueue;
import dataStructures.priorityQueue.PriorityQueue;
import dataStructures.tuple.Tuple2;

public class Huffman {

    // Exercise 1
    public static Dictionary<Character, Integer> weights(String s) {
        Dictionary<Character, Integer> res = new AVLDictionary<>();
        for (int i = 0; i < s.length(); ++i){
            if (res.isDefinedAt(s.charAt(i))) {
                res.insert(s.charAt(i), res.valueOf(s.charAt(i)) + 1);
            } else {
                res.insert(s.charAt(i), 1);
            }
        }
        return res;
    }

    // Exercise 2.a
    public static PriorityQueue<WLeafTree<Character>> huffmanLeaves(String s) {
    	PriorityQueue<WLeafTree<Character>> res = new BinaryHeapPriorityQueue<>();
        Dictionary<Character, Integer> w = weights(s);
        for (Character c: w.keys()) {
            res.enqueue(new WLeafTree(c, w.valueOf(c)));
        }
        return res;
    }

    // Exercise 2.b
    public static WLeafTree<Character> huffmanTree(String s) {
    	if (weights(s).size() < 2)
            throw new HuffmanException("Hay menos de dos caracteres distintos");
        PriorityQueue<WLeafTree<Character>> pq = huffmanLeaves(s);
        return huffmanTreeAux(pq.first(), huffmanLeaves(s));
    }

    public static WLeafTree<Character> huffmanTreeAux (WLeafTree<Character> first, PriorityQueue<WLeafTree<Character>> pq) {
        pq.dequeue();
        if (pq.isEmpty()) {
            return first;
        } else {
            return huffmanTreeAux(merge (first, pq.first()), pq);
        }
    }

    public static WLeafTree<Character> merge (WLeafTree<Character> left, WLeafTree<Character> right) {
        return new WLeafTree<>(left, right);
    }

    // Exercise 3.a
    public static Dictionary<Character, List<Integer>> joinDics(Dictionary<Character, List<Integer>> d1, Dictionary<Character, List<Integer>> d2) {
        Dictionary<Character, List<Integer>> res = new AVLDictionary<>();
        for (Tuple2<Character,List<Integer>> t : d1.keysValues()){
            res.insert(t._1(), t._2());
        }
        for (Tuple2<Character,List<Integer>> t : d2.keysValues()){
            res.insert(t._1(), t._2());
        }
        return res;
    }

    // Exercise 3.b
    public static Dictionary<Character, List<Integer>> prefixWith(int i, Dictionary<Character, List<Integer>> d) {
        Dictionary<Character, List<Integer>> res = new AVLDictionary<>();
        for (Tuple2<Character,List<Integer>> t : d.keysValues()){
            List<Integer> l2 = t._2();
            l2.insert(0, i);
            res.insert(t._1(),t._2());
        }
        return res;
    }

    // Exercise 3.c
    public static Dictionary<Character, List<Integer>> huffmanCode(WLeafTree<Character> ht) {
        Dictionary<Character, List<Integer>> res = new AVLDictionary<>();
        if (ht.isLeaf()){
            res.insert(ht.elem(), new LinkedList<>());
        } else {
            Dictionary<Character, List<Integer>> d1 = prefixWith(0, huffmanCode(ht.leftChild()));
            Dictionary<Character, List<Integer>> d2 = prefixWith(1, huffmanCode(ht.rightChild()));
            res = joinDics(d1, d2);
        }
        return res;
    }

    // Exercise 4
    public static List<Integer> encode(String s, Dictionary<Character, List<Integer>> hc) {
        //to do 
    	return null;
    }

    // Exercise 5
    public static String decode(List<Integer> bits, WLeafTree<Character> ht) {
        //to do 
    	return null;
    }
}

File: Exámenes/2016 SEPT/Intento 1/haskell/Huffman.hs
-- | Data Structures
-- | September, 2016
-- |
-- | Student's name: Ángel Manuel Soria Gil
-- | Student's group: Doble Grado Matemáticas e igeniería informática

module Huffman where

import qualified DataStructures.Dictionary.AVLDictionary as D
import qualified DataStructures.PriorityQueue.WBLeftistHeapPriorityQueue as PQ
import Data.List (nub)

-- | Exercise 1

weights :: Ord a => [a] -> D.Dictionary a Int
weights [] = D.empty
weights (x:xs) = D.updateOrInsert x f 1 (weights xs)
    where
        f v = v + 1
    

{-

> weights "abracadabra"
AVLDictionary('a'->5,'b'->2,'c'->1,'d'->1,'r'->2)

> weights [1,2,9,2,0,1,6,1,5,5,8]
AVLDictionary(0->1,1->3,2->2,5->2,6->1,8->1,9->1)

> weights ""
AVLDictionary()

-}


-- Implementation of Huffman Trees
data WLeafTree a = WLeaf a Int  -- Stored value (type a) and weight (type Int)
                 | WNode (WLeafTree a) (WLeafTree a) Int -- Left child, right child and weight
                 deriving (Eq, Show)

weight :: WLeafTree a -> Int
weight (WLeaf _ n)   = n
weight (WNode _ _ n) = n

-- Define order on trees according to their weights
instance Eq a => Ord (WLeafTree a) where
  wlt <= wlt' =  weight wlt <= weight wlt'

-- Build a new tree by joining two existing trees
merge :: WLeafTree a -> WLeafTree a -> WLeafTree a
merge wlt1 wlt2 = WNode wlt1 wlt2 (weight wlt1 + weight wlt2)


-- | Exercise 2

-- 2.a
huffmanLeaves :: String -> PQ.PQueue (WLeafTree Char)
huffmanLeaves str = huffmanLeavesAux (D.keysValues (weights str)) (PQ.empty)

huffmanLeavesAux :: [(Char, Int)] -> PQ.PQueue (WLeafTree Char) -> PQ.PQueue (WLeafTree Char)
huffmanLeavesAux [] res = res
huffmanLeavesAux ((x,y):xs) pq = huffmanLeavesAux xs (PQ.enqueue (WLeaf x y) pq)

{-

> huffmanLeaves "abracadabra"
WBLeftistHeapPriorityQueue(WLeaf 'c' 1,WLeaf 'd' 1,WLeaf 'b' 2,WLeaf 'r' 2,WLeaf 'a' 5)

-}

-- 2.b
huffmanTree :: String -> WLeafTree Char
huffmanTree str
    | n < 2 = error "the string must have at least two different symbols"
    | otherwise = huffmanTreeAux (huffmanLeaves str) n
    where n = D.size (weights str)

huffmanTreeAux :: PQ.PQueue (WLeafTree Char) -> Int -> WLeafTree Char
huffmanTreeAux pq n
    | n == 2 = solBase
    | otherwise = huffmanTreeAux newpq (n - 1)
    where
        solBase = merge (PQ.first pq) (PQ.first (PQ.dequeue pq))
        newpq = PQ.enqueue solBase (PQ.dequeue (PQ.dequeue pq))

{-

> printWLeafTree $ huffmanTree "abracadabra"
      11_______
      /        \
('a',5)        6_______________
              /                \
        ('r',2)          _______4
                        /       \
                       2        ('b',2)
                      / \
                ('c',1) ('d',1)

> printWLeafTree $ huffmanTree "abracadabra pata de cabra"
                         ______________________25_____________
                        /                                     \
         ______________10______                          ______15
        /                      \                        /       \
       4_______                6_______                6        ('a',9)
      /        \              /        \              / \
('d',2)        2        (' ',3)        3        ('b',3) ('r',3)
              / \                     / \
        ('e',1) ('p',1)         ('t',1) ('c',2)

> printWLeafTree $ huffmanTree "aaa"
*** Exception: huffmanTree: the string must have at least two different symbols

-}


-- | Exercise 3

-- 3.a
-- There is no key defined in both dictionaries
joinDics :: Ord a => D.Dictionary a b -> D.Dictionary a b -> D.Dictionary a b 
joinDics dict1 dict2 = joinDicsAux (D.keysValues dict1) dict2

joinDicsAux :: Ord a => [(a,b)] -> D.Dictionary a b -> D.Dictionary a b 
joinDicsAux [] dict = dict
joinDicsAux ((x,y):xs) dict = joinDicsAux xs (D.insert x y dict)

{-

> joinDics (D.insert 'a' 1 $ D.insert 'c' 3 $ D.empty) D.empty
AVLDictionary('a'->1,'c'->3)

> joinDics (D.insert 'a' 1 $ D.insert 'c' 3 $ D.empty) (D.insert 'b' 2 $ D.insert 'd' 4 $ D.insert 'e' 5 $ D.empty)
AVLDictionary('a'->1,'b'->2,'c'->3,'d'->4,'e'->5)

-}

-- 3.b
prefixWith :: Ord a => b -> D.Dictionary a [b] -> D.Dictionary a [b]
prefixWith x dict = prefixWithAux x (D.keys dict) dict

prefixWithAux :: Ord a => b -> [a] -> D.Dictionary a [b] -> D.Dictionary a [b]
prefixWithAux x [] dict = dict
prefixWithAux x (y:ys) dict = prefixWithAux x ys (D.updateOrInsert y f value dict)
    where
        f list = x:list
        value = case D.valueOf y dict of
            Nothing -> []
            Just list -> list

{-

> prefixWith 0 (D.insert 'a' [0,0,1] $ D.insert 'b' [1,0,0] $ D.empty)
AVLDictionary('a'->[0,0,0,1],'b'->[0,1,0,0])

> prefixWith 'h' (D.insert 1 "asta" $ D.insert 2 "echo" $ D.empty)
AVLDictionary(1->"hasta",2->"hecho")

-}

-- 3.c
huffmanCode :: WLeafTree Char -> D.Dictionary Char [Integer]
huffmanCode tree = huffmanCodeAux tree D.empty


huffmanCodeAux :: WLeafTree Char -> D.Dictionary Char [Integer] -> D.Dictionary Char [Integer]
huffmanCodeAux (WLeaf x n) dict = D.insert x [] dict
huffmanCodeAux (WNode lt rt n) dict = joinDics (prefixWith 0 (huffmanCodeAux lt dict)) (prefixWith 1 (huffmanCodeAux rt dict))


{-

> huffmanCode (huffmanTree "abracadabra")
AVLDictionary('a'->[0],'b'->[1,1,1],'c'->[1,1,0,0],'d'->[1,1,0,1],'r'->[1,0])

-}

-- ONLY for students not taking continuous assessment

-- | Exercise 4

encode :: String -> D.Dictionary Char [Integer] -> [Integer]
encode = undefined

{-

> encode "abracadabra" (huffmanCode (huffmanTree "abracadabra"))
[0,1,1,1,1,0,0,1,1,0,0,0,1,1,0,1,0,1,1,1,1,0,0]

-}

-- | Exercise 5

-- 5.a
takeSymbol :: [Integer] -> WLeafTree Char -> (Char, [Integer])
takeSymbol = undefined

{-

> takeSymbol [0,1,1,1,1,0,0,1,1,0,0,0,1,1,0,1,0,1,1,1,1,0,0] (huffmanTree "abracadabra")
('a',[1,1,1,1,0,0,1,1,0,0,0,1,1,0,1,0,1,1,1,1,0,0])

> takeSymbol [1,1,1,1,0,0,1,1,0,0,0,1,1,0,1,0,1,1,1,1,0,0] (huffmanTree "abracadabra")
('b',[1,0,0,1,1,0,0,0,1,1,0,1,0,1,1,1,1,0,0])

-}

-- 5.b
decode :: [Integer] -> WLeafTree Char -> String
decode = undefined

{-

> decode [0,1,1,1,1,0,0,1,1,0,0,0,1,1,0,1,0,1,1,1,1,0,0] (huffmanTree "abracadabra")
"abracadabra"

-}

-------------------------------------------------------------------------------
-- Pretty Printing a WLeafTree
-- (adapted from http://stackoverflow.com/questions/1733311/pretty-print-a-tree)
-------------------------------------------------------------------------------

printWLeafTree :: (Show a) => WLeafTree a -> IO ()
printWLeafTree t = putStrLn (unlines xss)
 where
   (xss,_,_,_) = pprint t

pprint :: Show a => WLeafTree a -> ([String], Int, Int, Int)
pprint (WLeaf x we)             =  ([s], ls, 0, ls-1)
  where
    s = show (x,we)
    ls = length s
pprint (WNode lt rt we)         =  (resultLines, w, lw'-swl, totLW+1+swr)
  where
    nSpaces n = replicate n ' '
    nBars n = replicate n '_'
    -- compute info for string of this node's data
    s = show we
    sw = length s
    swl = div sw 2
    swr = div (sw-1) 2
    (lp,lw,_,lc) = pprint lt
    (rp,rw,rc,_) = pprint rt
    -- recurse
    (lw',lb) = if lw==0 then (1," ") else (lw,"/")
    (rw',rb) = if rw==0 then (1," ") else (rw,"\\")
    -- compute full width of this tree
    totLW = maximum [lw', swl,  1]
    totRW = maximum [rw', swr, 1]
    w = totLW + 1 + totRW
{-
A suggestive example:
     dddd | d | dddd__
        / |   |       \
      lll |   |       rr
          |   |      ...
          |   | rrrrrrrrrrr
     ----       ----           swl, swr (left/right string width (of this node) before any padding)
      ---       -----------    lw, rw   (left/right width (of subtree) before any padding)
     ----                      totLW
                -----------    totRW
     ----   -   -----------    w (total width)
-}
    -- get right column info that accounts for left side
    rc2 = totLW + 1 + rc
    -- make left and right tree same height
    llp = length lp
    lrp = length rp
    lp' = if llp < lrp then lp ++ replicate (lrp - llp) "" else lp
    rp' = if lrp < llp then rp ++ replicate (llp - lrp) "" else rp
    -- widen left and right trees if necessary (in case parent node is wider, and also to fix the 'added height')
    lp'' = map (\s -> if length s < totLW then nSpaces (totLW - length s) ++ s else s) lp'
    rp'' = map (\s -> if length s < totRW then s ++ nSpaces (totRW - length s) else s) rp'
    -- first part of line1
    line1 = if swl < lw' - lc - 1 then
                nSpaces (lc + 1) ++ nBars (lw' - lc - swl) ++ s
            else
                nSpaces (totLW - swl) ++ s
    -- line1 right bars
    lline1 = length line1
    line1' = if rc2 > lline1 then
                line1 ++ nBars (rc2 - lline1)
             else
                line1
    -- line1 right padding
    line1'' = line1' ++ nSpaces (w - length line1')
    -- first part of line2
    line2 = nSpaces (totLW - lw' + lc) ++ lb
    -- pad rest of left half
    line2' = line2 ++ nSpaces (totLW - length line2)
    -- add right content
    line2'' = line2' ++ " " ++ nSpaces rc ++ rb
    -- add right padding
    line2''' = line2'' ++ nSpaces (w - length line2'')
    resultLines = line1'' : line2''' : zipWith (\lt rt -> lt ++ " " ++ rt) lp'' rp''
